### ISSUE TITLE
RuntimeException when creating service FirebaseReceiver

### ISSUE THREAD

**Issue by marstr** [Azure/azure-notificationhubs-android#123]
_Originally posted by @acastr7 in https://github.com/Azure/azure-notificationhubs-android/issues/116#issuecomment-676735132_

@marstr After updating to this version on Xamarin.Android I noticed our push notifications no longer work.

Here is the error we are seeing. It only happens on first launch and then we never see the FirebaseReceiver start again:
```
[AndroidRuntime] java.lang.RuntimeException: Unable to create service com.microsoft.windowsazure.messaging.notificationhubs.FirebaseReceiver: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.microsoft.windowsazure.messaging.notificationhubs.PushChannelVisitor.getPushChannel()' on a null object reference
[AndroidRuntime] 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3965)
[AndroidRuntime] 	at android.app.ActivityThread.access$1500(ActivityThread.java:219)
[AndroidRuntime] 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1875)
[AndroidRuntime] 	at android.os.Handler.dispatchMessage(Handler.java:107)
[AndroidRuntime] 	at a08-19 17:28:05.991 E/AndroidRuntime( 9209): 	at android.os.Looper.loop(Looper.java:214)
[AndroidRuntime] 	at android.app.ActivityThread.main(ActivityThread.java:7356)
[AndroidRuntime] 	at java.lang.reflect.Method.invoke(Native Method)
[AndroidRuntime] 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
[AndroidRuntime] 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
[AndroidRuntime] Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.microsoft.windowsazure.messaging.notificationhubs.PushChannelVisitor.getPushChannel()' on a null object reference
[AndroidRuntime] 	at com.microsoft.windowsazure.messaging.notificationhubs.NotificationHub.getP08-19 17:28:05.991 E/AndroidRuntime( 9209): 	at com.microsoft.windowsazure.messaging.notificationhubs.NotificationHub.getInstancePushChannel(NotificationHub.java:257)
[AndroidRuntime] 	at com.microsoft.windowsazure.messaging.notificationhubs.FirebaseReceiver.onCreate(FirebaseReceiver.java:50)
[AndroidRuntime] 	at android.app.ActivityThread.handleCreateService(ActivityThread.java:3953)
[AndroidRuntime] 	... 8 more
```

https://github.com/Azure/azure-notificationhubs-android/blob/4dc585a982854bc7ab7ac8eab0ef9fc7ecec731e/notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/FirebaseReceiver.java#L29-L31

This creates a NotificationHub with a null PushChannelVisitor so when it tries to get the channel it crashes on create. I'm thinking the NotificationHub class needs a `IsInitalized` flag that gets set on `NotificationHub.Start` and that gets checked inside the `FirebaseReceiver` before every call to the notification hub.

Also our app does not call `NotifcationHub.Start` until after the user has logged in since our hub connection strings is api driven. I feel like the following code below only needs to be called after `NotificationHub.Start` is called, not with the firebase service is booted up.

https://github.com/Azure/azure-notificationhubs-android/blob/4dc585a982854bc7ab7ac8eab0ef9fc7ecec731e/notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/FirebaseReceiver.java#L50-L62

**Comment by marstr** [comment]
Thanks for this report, @acastr7. It's brought my attention to a race condition in initialization and we need to fix it. Unfortunately, I don't think it's caused by the new version, I think you're just happening to hit the race condition now.

The is initialized logic seems like it would duplicate our "enabled" mechanism. I'll submit a patch to fix this ASAP, and include it with the v1.1.1 release.

**Comment by marstr** [comment]
While working on the PR I just submitted, I was able to reproduce your problem consistently by commenting out most things in the sample app and not letting `NotificationHub.start()` get called. Refactoring out the injection of the `Application` context so that either application code or `FirebaseReceiver` can win the race should fix your particular situation.

I'm going to have to keep thinking about your second point of waiting to get the Firebase InstanceId. On one hand, it shouldn't make too much difference because in theory the Instance ID won't change before or after your user's logged in. But I do think this has exposed a fragility that needs to be addressed. For example, if a user deletes application data, would Firebase generate a new token? If not, then we would presumably lose the one stored in SharedPreferences nor would we get a callback for a new token. At that point we're in big trouble, and the system will likely never bother starting FirebaseReceiver again.

I think instead of just delaying fetching the FirebaseInstanceId, we should pull out a separate interface for fetching the PushChannel from whatever the push provider is, and be able to do that at will, potentially whenever we create a new Installation record for safety.
